File: scripts/test_solutions/329_longest_increasing_path_matrix.py
def longestIncreasingPath(matrix):
    m, n = len(matrix), len(matrix[0])
    memo = [[0] * n for _ in range(m)]

    def dfs(i, j):
        if memo[i][j]:
            return memo[i][j]
        max_len = 1
        for di, dj in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
            ni, nj = i + di, j + dj
            if 0 <= ni < m and 0 <= nj < n and matrix[ni][nj] > matrix[i][j]:
                max_len = max(max_len, 1 + dfs(ni, nj))
        memo[i][j] = max_len
        return max_len

    result = 0
    for i in range(m):
        for j in range(n):
            result = max(result, dfs(i, j))
    return result

File: scripts/test_solutions/329_longest_increasing_path_matrix.cpp
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        memo.assign(m, vector<int>(n, 0));

        int result = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                result = max(result, dfs(matrix, i, j));
            }
        }
        return result;
    }

private:
    vector<vector<int>> memo;
    int dirs[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    int dfs(vector<vector<int>>& matrix, int i, int j) {
        if (memo[i][j]) return memo[i][j];
        int m = matrix.size(), n = matrix[0].size();
        int maxLen = 1;
        for (auto& d : dirs) {
            int ni = i + d[0], nj = j + d[1];
            if (ni >= 0 && ni < m && nj >= 0 && nj < n && matrix[ni][nj] > matrix[i][j]) {
                maxLen = max(maxLen, 1 + dfs(matrix, ni, nj));
            }
        }
        memo[i][j] = maxLen;
        return maxLen;
    }
};
